tests/std/tests/P2502R2_generator/test.cpp

```diff
@@ -0,0 +1,429 @@
+    auto it2        = ranges::begin(r2);
+    const auto end1 = ranges::end(r1);
+    const auto end2 = ranges::end(r2);
+    for (; it1 != end1 && it2 != end2; ++it1, ++it2) {
+        co_yield {*it1, *it2};
+    }
+}
+
+void zip_example() {
+    using V = tuple<int, int>;
+    using R = tuple<int&, int&>;
+
+    auto g0 = co_zip(array{1, 2, 3}, vector{10, 20, 30, 40, 50});
+    test_one<gen_traits<R, V>, V, R, R>(move(g0), array{tuple{1, 10}, tuple{2, 20}, tuple{3, 30}});
+
+    g0 = co_zip(array{3, 2, 1}, vector{10, 20, 30, 40, 50});
+    test_one<gen_traits<R, V>, V, R, R>(move(g0), array{tuple{3, 10}, tuple{2, 20}, tuple{1, 30}});
+}
+
+template <class Reference = const int&>
+generator<Reference, int> co_upto(const int hi) {
+    assert(hi >= 0);
+    for (int i = 0; i < hi; ++i) {
+        co_yield i;
+    }
+}
+
+void test_weird_reference_types() {
+    constexpr int n = 32;
+    { // Test mutable lvalue reference type
+        auto r   = co_upto<int&>(n);
+        auto pos = r.begin();
+        for (int i = 0; i < n / 2; ++i, ++*pos, ++pos) {
+            assert(pos != r.end());
+            assert(*pos == 2 * i);
+        }
+        assert(pos == r.end());
+    }
+
+#if !(defined(__clang__) && defined(_M_IX86)) // TRANSITION, LLVM-56507
+    { // Test with mutable xvalue reference type
+        auto woof = [](size_t size, size_t count) -> generator<vector<int>&&> {
+            random_device rd{};
+            uniform_int_distribution dist{0, 99};
+            vector<int> vec;
+            while (count-- > 0) {
+                vec.resize(size);
+                ranges::generate(vec, [&] { return dist(rd); });
+                co_yield move(vec);
+                assert(vec.empty()); // when we yield an rvalue, the caller moves from it
+            }
+
+            // Test yielding lvalue
+            vec.resize(size);
+            ranges::generate(vec, [&] { return dist(rd); });
+            const auto tmp = vec;
+            co_yield vec;
+            assert(tmp == vec); // when we yield an lvalue, the caller moves from a copy
+        };
+
+        constexpr size_t size = 16;
+        auto r                = woof(size, 4);
+        for (auto i = r.begin(); i != r.end(); ++i) {
+            vector<int> vec = *i;
+            assert(vec.size() == size);
+        }
+    }
+#endif // ^^^ no workaround ^^^
+}
+
+#if !(defined(__clang__) && defined(_M_IX86)) // TRANSITION, LLVM-56507
+generator<int> iota_repeater(const int hi, const int depth) {
+    if (depth > 0) {
+        co_yield ranges::elements_of(iota_repeater(hi, depth - 1));
+        co_yield ranges::elements_of(iota_repeater(hi, depth - 1));
+    } else {
+        co_yield ranges::elements_of(co_upto<int>(hi));
+    }
+}
+
+void recursive_test() {
+    static constexpr auto might_throw = []() -> generator<int> {
+        co_yield 0;
+        throw runtime_error{"error"};
+    };
+
+    static constexpr auto nested_ints = []() -> generator<int> {

```

Comment:
As scalars, these don't need to be `static`, they can be plain `constexpr`.

Reply:
Ah, the second has to capture the first by value. That's a bigger change than I thought (the lambda is no longer empty) but I think making the dependency more explicit is a win.